contents/c++/rvalue.md

```diff
@@ -0,0 +1,169 @@
+
+However, Value Semantics has [one major flaw](https://www.quora.com/What-are-the-drawbacks-of-pass-by-value-result):
+* Poor performance and scalability. If the function called is a read-only function, creating an additional copy will be an unnecessary consumption of memory and runtime. This is especially true when scaling up to pass around objects of larger size, as more memory and runtime is needed to create each copy.
+
+### Reference Semantics
+Reference (or pointer) semantics is another choice available for use in C++. C++ allows users to declare [pointers](http://www.cplusplus.com/doc/tutorial/pointers/) and [references](https://www.geeksforgeeks.org/references-in-c/) that point to the address of the object. As such, we can pass around these reference address to functions and declaration of other pointers, and all of them will refer to and use the same object and address.
+
+Some advantages of Reference Semantics include:
+* Improve performance and scalability. By passing around pointers and references to the same object, functions can just use the values in that object directly without having to create an extra copy, thus overcoming the major flaw of Value Semantics.
+* Allows interaction between objects and functions. Since the same object can be referred to by the different pointers, they can modify the same address values in different areas.
+
+However, Reference Semantics also comes with its own problems, which include:
+* Indeterminant behaviour in a multi-threaded environment. Since each thread will all be pointing to the same object, this will lead to unnecessary data races and indeterminant behaviour as these same objects are modified concurrently in each thread. This in turn requires extra work to synchronize the values between objects, which also slows down the program. More details of this can be found in the resources [Java Concurrency](java/JavaConcurrency.md) and [Java Synchronization](java/JavaSynchronization.md).
+* Unintended behaviour. If users are not careful, they can modify the values in the object using one pointer or reference, which will then be reflected by another pointer to the same object, which the user may not have intended. A good example of this problem can be found in this [article](http://www.drdobbs.com/cpp/optimization-calling-by-value-or-by-refe/232400151).
+
+## Move Semantics
+From the summary above, there are many benefits for using Value Semantics. However, it's one major flaw is that a copy of the object will always be created, which can computationally expensive if this object is of a large size. However, Reference Semantics may also not be preferred due to its different problems as discussed above. As such, in order to continue gaining the benefits of Value Semantics while overcoming it's major flaw, `C++11` introduced a new mode to users, Move Semantics.
```

Comment:
This paragraph has sentences that are fragmented. 

For instance, you briefly mention that Reference Semantics has flaws and then move on to say Move Semantics mitigates the flaws of Value Semantics without linking it to Reference Semantics. 

So i'm not sure what this paragraph is about. Are you trying to say the Move Semantics mitigate the flaws of both Value and Reference Semantics or just Value Semantics?

Reply:
Hmmm, I'm kind of trying to say that we might want to continue using Value Semantics because of its benefits, and Move Semantics help to mitigate its major flaw.
Perhaps I should not even mention Reference Semantics here?

Reply:
Would this sound better:
`As such, in order to continue gaining the benefits Value Semantics have over Reference Semantics while overcoming it's major flaw, `C++11` introduced a new mode to users, Move Semantics.`

Reply:
Yup, this sounds better